packages/styled-base/types/index.d.ts

```diff
@@ -24,127 +23,88 @@ export {
-  ): StyledComponent<InnerProps, StyleProps, StyledInstanceTheme>
-  <
-    StyleProps extends Omit<
-      Overwrapped<InnerProps, StyleProps>,
-      ReactClassPropKeys
-    > = Omit<InnerProps & ExtraProps, ReactClassPropKeys>
-  >(
+  /**
+   * @typeparam AdditionalProps  Additional props to add to your styled component
+   */
+  <AdditionalProps extends object>(
+    ...styles: Array<
+      Interpolation<
+        ComponentProps & SpecificComponentProps & StyleProps & AdditionalProps
+      >
+    >
+  ): StyledComponent<ComponentProps & AdditionalProps, SpecificComponentProps>
+  /**
+   * @typeparam AdditionalProps  Additional props to add to your styled component
+   */
+  <AdditionalProps extends object>(
     template: TemplateStringsArray,
-    ...styles: Array<Interpolation<WithTheme<StyleProps, StyledInstanceTheme>>>
-  ): StyledComponent<InnerProps, StyleProps, StyledInstanceTheme>
-}
-
-export type CreateStyledComponentBase<
-  InnerProps,
-  ExtraProps,
-  StyledInstanceTheme extends object
-> =
-  // this "reversed" condition checks if StyledInstanceTheme was already parametrized when using CreateStyled
-  object extends StyledInstanceTheme
-    ? CreateStyledComponentBaseThemeless<InnerProps, ExtraProps>
-    : CreateStyledComponentBaseThemed<
-        InnerProps,
-        ExtraProps,
-        StyledInstanceTheme
+    ...styles: Array<
+      Interpolation<
+        ComponentProps &
+          SpecificComponentProps &
+          AdditionalProps &
+          AdditionalProps
       >
-
-export type CreateStyledComponentIntrinsic<
-  Tag extends keyof JSXInEl,
-  ExtraProps,
-  Theme extends object
-> = CreateStyledComponentBase<JSXInEl[Tag], ExtraProps, Theme>
-export type CreateStyledComponentExtrinsic<
-  Tag extends React.ComponentType<any>,
-  ExtraProps,
-  Theme extends object
-> = CreateStyledComponentBase<PropsOf<Tag>, ExtraProps, Theme>
+    >
+  ): StyledComponent<ComponentProps & AdditionalProps, SpecificComponentProps>
+}
 
 /**
  * @desc
- * This function accepts `InnerProps`/`Tag` to infer the type of `tag`,
- * and accepts `ExtraProps` for user who use string style
- * to be able to declare extra props without using
- * `` styled('button')<ExtraProps>`...` ``, which does not supported in
- * styled-component VSCode extension.
- * If your tool support syntax highlighting for `` styled('button')<ExtraProps>`...` ``
- * it could be more efficient.
+ * This function accepts a React component or tag ('div', 'a' etc).
+ *
+ * @example styled(MyComponent)({ width: 100 })
+ * @example styled(MyComponent)(myComponentProps => ({ width: myComponentProps.width })
+ * @example styled('div')({ width: 100 })
+ * @example styled('div')<Props>(props => ({ width: props.width })
  */
-export interface CreateStyled<Theme extends object = any> {
-  <Tag extends React.ComponentType<any>, ExtraProps = {}>(
-    tag: Tag,
+export interface CreateStyled<Theme = any> {
```

Comment:
I think I've mentioned this elsewhere, but better to repeat myself than to miss this. 

This should both restrict Theme to be `extends {}` and provide a `{}` as default rather than `any`

Reply:
In this case I think `any` is correct. The other instance I need to check. `Theme` is still used as the type for the `theme` key.

Reply:
Or does emotion restrict theme to be an object?

Reply:
Going to restore the `extends object` though

Reply:
As mentioned elsewhere - we kinda assume in runtime this is an object (and provide default), so I think `extends {}` would be correct. If you don't agree - let's discuss more, I might be wrong.

Reply:
Done!